Userland/Libraries/LibWeb/WebSockets/WebSocket.cpp

```diff
@@ -50,50 +51,107 @@ WebSocketClientManager::WebSocketClientManager() = default;
-                if (character < '\x21' || character > '\x7E')
-                    return WebIDL::SyntaxError::create(realm, "Found invalid character in subprotocol name");
-            }
+    // 8. If protocols is a string, set protocols to a sequence consisting of just that string.
+    if (protocols.has_value() && protocols->has<String>())
+        protocols_sequence = { protocols.value().get<String>() };
+    else if (protocols.has_value() && protocols->has<Vector<String>>())
+        protocols_sequence = protocols.value().get<Vector<String>>();
+    else
+        protocols_sequence = {};
+
+    // 9. If any of the values in protocols occur more than once or otherwise fail to match the requirements for elements that comprise
+    //    the value of `Sec-WebSocket-Protocol` fields as defined by The WebSocket protocol, then throw a "SyntaxError" DOMException. [WSP]
+    auto sorted_protocols = protocols_sequence;
+    quick_sort(sorted_protocols);
+    for (size_t i = 0; i < sorted_protocols.size(); i++) {
+        // https://datatracker.ietf.org/doc/html/rfc6455
+        // The elements that comprise this value MUST be non-empty strings with characters in the range U+0021 to U+007E not including
+        // separator characters as defined in [RFC2616] and MUST all be unique strings.
+        auto protocol = sorted_protocols[i];
+        if (i < sorted_protocols.size() - 1 && protocol == sorted_protocols[i + 1])
+            return WebIDL::SyntaxError::create(realm, "Found a duplicate protocol name in the specified list"sv);
+        for (auto code_point : protocol.code_points()) {
+            if (code_point < '\x21' || code_point > '\x7E')
+                return WebIDL::SyntaxError::create(realm, "Found invalid character in subprotocol name"sv);
         }
     }
-    return MUST_OR_THROW_OOM(realm.heap().allocate<WebSocket>(realm, window, url_record, protocols_sequence));
+
+    // 10. Set this's url to urlRecord.
+    web_socket->set_url(url_record);
+
+    // 11. Let client be this’s relevant settings object.
+    auto& client = relevant_settings_object;
+
+    // FIXME: 12. Run this step in parallel:
+    //     1. Establish a WebSocket connection given urlRecord, protocols, and client. [FETCH]
+    TRY_OR_THROW_OOM(vm, web_socket->establish_web_socket_connection(url_record, protocols_sequence, client));
+
+    return web_socket;
+}
+
+WebSocket::WebSocket(JS::Realm& realm)
+    : EventTarget(realm)
+{
 }
 
-WebSocket::WebSocket(HTML::Window& window, AK::URL& url, Vector<String> const& protocols)
-    : EventTarget(window.realm())
-    , m_window(window)
+WebSocket::~WebSocket() = default;
+
+JS::ThrowCompletionOr<void> WebSocket::initialize(JS::Realm& realm)
+{
+    MUST_OR_THROW_OOM(Base::initialize(realm));
+    set_prototype(&Bindings::ensure_web_prototype<Bindings::WebSocketPrototype>(realm, "WebSocket"));
+
+    return {};
+}
+
+void WebSocket::visit_edges(Cell::Visitor& visitor)
```

Comment:
I think we could remove this function now now that we don't visit anything

Reply:
Thanks, removed it